llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td

```diff
@@ -435,6 +435,80 @@ def IREELinalgExt_TopkOp : IREELinalgExt_Op<"topk",[
   }];
 }
 
+// TODO: is pack the right name? After all we are re-arranging the tensor.
+// TODO: destination passing style.
+def IREELinalgExt_PackOp : IREELinalgExt_Op<"pack",[
+  DeclareOpInterfaceMethods<LinalgExtInterface>,
+  DeclareOpInterfaceMethods<TilingInterface,
+    ["getIterationDomain",
+     "getLoopIteratorTypes",
+     "generateScalarImplementation"]>,
+  DeclareOpInterfaceMethods<MemoryEffectsOpInterface>
+]>{
+  let summary = "pack operation"; 
+
+  let arguments = (ins Variadic<AnyShaped>:$inputs,
+                       Variadic<AnyShaped>:$outputs,
+                       OptionalAttr<DefaultValuedAttr<I64ArrayAttr, "{}">>:$iterator_interchange,
+                       Variadic<Index>:$inner_tiles,
+                       I64ArrayAttr:$static_inner_tiles);
+
+  let results = (outs Variadic<AnyRankedTensor>:$results);
+  let assemblyFormat = [{
+    attr-dict
+    $inputs 
+    `inner_tiles`
+    custom<DynamicIndexList>($inner_tiles, $static_inner_tiles,
+                             "ShapedType::kDynamicStrideOrOffset")
+    (`interchange` `=` $iterator_interchange^)?
+    `into` $outputs `:` `(` type($inputs) type($outputs) `)`
+     (`->` type($results)^)?
+  }];
+
+  let extraClassDeclaration = extraLinalgExtOpClassDeclaration # [{
+
+    // Return the output operand `destination`.
+    Value getOutput() {
+      return getOutputOperand(0)->get();
+    }
+
+    // Return the input operand `source`.
+    Value getInput() {
+      return getInputOperand(0)->get();
+    }
+
+    // Return the output rank.
+    int64_t getOutputRank() {
+      return  getOutput().getType().cast<ShapedType>().getRank();
+    }
+
+    // Return the output type.
+    ShapedType getOutputType() {
+      return getOutput().getType();
+    }
+
+    // Return the output shape.
+    ArrayRef<int64_t> getOutputShape() {
+      return getOutputType().cast<ShapedType>().getShape();
+    }
+
+    // Return the rank of the input operand.
+    int64_t getInputRank() {
+      return getInput().getType().cast<ShapedType>().getRank();
+    }
+
+    // Return the tile sizes ad OpFoldResults.
+    SmallVector<OpFoldResult> getMixedInnerTiles();
+
+    // Infer and return the shape of the output.
+    SmallVector<OpFoldResult> buildOutputShape(OpBuilder &builder);
```

Comment:
I would rather implement this under the `ReifyRankedShapedTypeOpInterface`. (https://github.com/llvm/llvm-project/blob/0ca5993741877ab7fd27a251cbc1895bd092d5ee/mlir/include/mlir/Interfaces/InferTypeOpInterface.td#L202) . It doesnt return an `OpFoldResult` unfortunately (that was my plan to start with but there was push back). But you can use the conversions from `Value` to `OpFoldResult` to get by.